### PR TITLE
Allow for the TagHelperTypeDescriptorResolver to be injected

### PR DESCRIPTION
**Pull request by javiercn**
Changes required in Razor to support https://github.com/aspnet/Mvc/issues/4087
* Add an interface for ITagHelperTypeResolver and ITagHelperDescriptorFactory
* Extract the default convention for tag helper discovery into TagHelperConventions following the same  pattern we use for ViewComponents.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperConventions.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Razor.TagHelpers
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperConventions.cs]: `Microsoft.AspNetCore.Razor.Runtime.TagHelpers`

File: src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperConventions.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+{
+    /// <summary>
+    /// Default convention for determining if a type is a tag helper.
+    /// </summary>
+    public static class TagHelperConventions
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperConventions.cs]: Is this just for consistency with the `ViewComponent` system or is there another need for this?

Comment (javiercn) [inline src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperConventions.cs]: I need it for AddTagHelpersAsServices and we don't want to duplicate the logic for discovering tag helper types. We do the same thing in view components

Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperConventions.cs]: So this is preemptive for https://github.com/aspnet/Mvc/issues/4089 ?

Comment (javiercn) [inline src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperConventions.cs]: Yes, for the tag helper part of it

**Comment by NTaylorMullen** [comment]
:shipit:

**Pull request merged**